### PR TITLE
[docs] Add profiling docs

### PR DESCRIPTION
**Pull request by ananthsub**

### CODE REVIEW

File: docs/profiling.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
+# Profiling
+
+Megatron Bridge provides built-in support for profiling your training jobs using various performance analysis tools, including NVIDIA Nsight Systems (Nsys) for workflow optimization and PyTorch-based profilers and memory trackers for tracking performance and memory usage patterns during training.
+
+## ProfilingConfig Overview
+
+`ProfilingConfig` is a dataclass that encapsulates profiling-related settings for training. It resides inside the overall `ConfigContainer`, which represents the complete configuration for a training run. For the complete API and all available fields, see {py:class}`~megatron.bridge.training.config.ProfilingConfig`.
```
Comment (jgerh) [inline docs/profiling.md]: suggest adding a link to ProfilingConfig.

```suggestion
`ProfilingConfig` is a dataclass that encapsulates profiling-related settings for training. It resides inside the overall `ConfigContainer`, which represents the complete configuration for a training run. For the complete API and all available fields, see {py:class}`~[megatron.bridge.training.config.ProfilingConfig](https://docs.nvidia.com/nemo/megatron-bridge/latest/apidocs/bridge/bridge.training.config.html#bridge.training.config.ProfilingConfig)`.

```

File: docs/profiling.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
+# Profiling
+
+Megatron Bridge provides built-in support for profiling your training jobs using various performance analysis tools, including NVIDIA Nsight Systems (Nsys) for workflow optimization and PyTorch-based profilers and memory trackers for tracking performance and memory usage patterns during training.
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
Megatron Bridge provides built-in support for profiling training jobs using a range of performance analysis tools. These include NVIDIA Nsight Systems (Nsys) for workflow optimization, as well as PyTorch-based profilers and memory trackers to monitor performance and memory usage patterns during training.
```

File: docs/profiling.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+`ProfilingConfig` is a dataclass that encapsulates profiling-related settings for training. It resides inside the overall `ConfigContainer`, which represents the complete configuration for a training run. For the complete API and all available fields, see {py:class}`~megatron.bridge.training.config.ProfilingConfig`.
+
+### Profiling Selection
+
+The configuration provides mutually exclusive profiling options. You can enable either NSys profiling (`use_nsys_profiler`) or PyTorch profiling (`use_pytorch_profiler`), but not both simultaneously.
+
+### Step Range and Ranks
+
+All profiling modes support configuring the step range (`profile_step_start` and `profile_step_end`) and target ranks (`profile_ranks`). By default, profiling targets rank 0, but you can specify multiple ranks to profile different parts of your distributed training setup.
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion

### Step Range and Target Ranks

All profiling modes allow you to configure:

- **Step range**: `profile_step_start` and `profile_step_end`
- **Target ranks**: `profile_ranks`

By default, profiling targets rank 0. You can specify multiple ranks to analyze different parts of your distributed training setup.
```

File: docs/profiling.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+### Additional Features
+
+The configuration includes options for recording tensor shapes (`record_shapes`) and memory profiling (`record_memory_history` with configurable `memory_snapshot_path`). These features provide deeper insights into your model's memory usage patterns and tensor operations.
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion

### Advanced Profiling Features

The configuration includes options for recording tensor shapes (`record_shapes`) and enabling memory profiling (`record_memory_history`) with a customizable output path (`memory_snapshot_path`). These features offer deeper visibility into your model’s memory consumption and tensor-level operations during training.

```

File: docs/profiling.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+## NSys Profiling
+
+NVIDIA Nsys is a system-wide performance analysis tool designed to help you tune and optimize CUDA applications. Megatron Bridge integrates with Nsys to enable profiling specific steps of your training job, making it easy to collect detailed performance data without manual instrumentation.
+
+```{note}
+NSys profiling cannot be used with the `FaultTolerancePlugin` due to implementation conflicts. If both are enabled, the framework will automatically disable NSys profiling and emit a warning.
+```
+
+### Configuration Options
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
### Configure NSys Profiling
```

File: docs/profiling.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+cfg.profiling = ProfilingConfig(
+    use_nsys_profiler=True,
+    profile_step_start=10,
+    profile_step_end=15,
+    profile_ranks=[0, 1],  # Profile first two ranks
+    record_shapes=False,   # Optional: record tensor shapes
+)
+```
+
+### Launching with NSys
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
### Launch with NSys
```

File: docs/profiling.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+When using NSys profiling, launch your training script with the NSys command wrapper:
+
+```bash
+nsys profile -s none -o <profile_filepath> -t cuda,nvtx --force-overwrite true --capture-range=cudaProfilerApi --capture-range-end=stop python <path_to_script>
+```
+
+Replace `<profile_filepath>` with your desired output path and `<path_to_script>` with your training script. The `--capture-range=cudaProfilerApi` option ensures profiling is controlled by the framework's step range configuration.
+
+### NSys Plugin for NeMo-Run Users
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
### Configure Profiling with the NSys Plugin
```

File: docs/profiling.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+Recipe users can leverage the `NsysPlugin` to configure NSys profiling through NeMo-Run executors. The plugin provides a convenient interface for setting up profiling without manually configuring the underlying NSys command.
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
Recipe users can leverage the `NsysPlugin` to configure NSys profiling through NeMo Run executors. The plugin provides a convenient interface for setting up profiling without manually configuring the underlying NSys command.
```

File: docs/profiling.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+# Run with profiling enabled
+with run.Experiment("nsys_profiling_experiment") as exp:
+    exp.add(recipe, executor=executor, plugins=plugins)
+    exp.run()
+```
+
+The plugin automatically configures the NSys command line options and sets up the profiling configuration in your training job.
+
+### Analyzing Results
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
### Analyze Results
```

File: docs/profiling.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+Once your profiling run is complete, you'll have generated NSys profile files that can be opened with the NSys GUI. Install [NVIDIA Nsight Systems](https://developer.nvidia.com/nsight-systems) from the NVIDIA Developer website, open the generated `.nsys-rep` files, and use the timeline view to examine your training job's performance characteristics.
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
After your profiling run completes, the NSys profile files (`.nsys-rep`) will be generated. To analyze them, install [NVIDIA Nsight Systems](https://developer.nvidia.com/nsight-systems) from the NVIDIA Developer website, open the files in the NSys GUI, and use the timeline view to explore the performance characteristics of your training job.
```

File: docs/profiling.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+## PyTorch Profiler
+
+Megatron Bridge supports the built-in PyTorch profiler, which is useful for viewing profiles in TensorBoard and understanding PyTorch-level performance characteristics.
+
+### Configuration
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
### Configure PyTorch Profiler
```

File: docs/profiling.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+cfg.profiling = ProfilingConfig(
+    use_pytorch_profiler=True,
+    profile_step_start=10,
+    profile_step_end=15,
+    profile_ranks=[0],
+    record_shapes=True,    # Record tensor shapes for detailed analysis
+)
+```
+
+### PyTorch Profiler Plugin
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
### Configure Profiling with the PyTorch Profiler Plugin
```

File: docs/profiling.md [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+]
+```
+
+## Memory Profiling
+
+Megatron Bridge provides built-in support for CUDA memory profiling to track and analyze memory usage patterns during training, including GPU memory allocation and consumption tracking.
+
+More information about the generated memory profiles can be found [here](https://pytorch.org/blog/understanding-gpu-memory-1/).
+
+### Configuration
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
### Configure Memory Profiling
```

File: docs/profiling.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    use_pytorch_profiler=True,  # or use_nsys_profiler=True
+    profile_step_start=10,
+    profile_step_end=15,
+    profile_ranks=[0],
+    record_memory_history=True,
+    memory_snapshot_path="memory_trace.pickle",  # Customize output path
+)
+```
+
+### Memory Analysis
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
### Analyze Memory Usage
```

File: docs/profiling.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+Once the run completes, the specified path will contain memory snapshots for each specified rank. These traces can be loaded with the PyTorch Memory Viz tool to plot memory usage over time and identify memory bottlenecks or leaks in your training pipeline.
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
After the run completes, memory snapshots for each specified rank are saved to the designated path. Load these traces using the PyTorch Memory Viz tool to plot memory usage over time and detect bottlenecks or leaks in your training pipeline.
```

File: docs/profiling.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+## Performance Considerations
```
Comment (jgerh) [inline docs/profiling.md]:
```suggestion
## Optimize Profiling Accuracy
```

File: docs/profiling.md [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+Megatron Bridge provides built-in support for profiling your training jobs using various performance analysis tools, including NVIDIA Nsight Systems (Nsys) for workflow optimization and PyTorch-based profilers and memory trackers for tracking performance and memory usage patterns during training.
+
+## ProfilingConfig Overview
+
+`ProfilingConfig` is a dataclass that encapsulates profiling-related settings for training. It resides inside the overall `ConfigContainer`, which represents the complete configuration for a training run. For the complete API and all available fields, see {py:class}`~megatron.bridge.training.config.ProfilingConfig`.
+
+### Profiling Selection
+
+The configuration provides mutually exclusive profiling options. You can enable either NSys profiling (`use_nsys_profiler`) or PyTorch profiling (`use_pytorch_profiler`), but not both simultaneously.
```
Comment (jgerh) [inline docs/profiling.md]: suggest different presentation for improved readability:
```suggestion
### Profiling Options

The configuration supports two mutually exclusive profiling options:

- **NSys profiling** (`use_nsys_profiler`)
- **PyTorch profiling** (`use_pytorch_profiler`)

You can enable one or the other, but not both at the same time.
```

**Review by jgerh — COMMENTED** [review]
Completed the tech pubs review of the file: suggested text edits to improve list item readability, revised headings for consistency across sections, and applied general copyedits.

File: docs/profiling.md [hunk 18]
```diff
@@ -0,0 +1,174 @@
+# Profiling
+
+Megatron Bridge provides built-in support for profiling training jobs using a range of performance analysis tools. These include NVIDIA Nsight Systems (Nsys) for workflow optimization, as well as PyTorch-based profilers and memory trackers to monitor performance and memory usage patterns during training.
+
+## ProfilingConfig Overview
+
+`ProfilingConfig` is a dataclass that encapsulates profiling-related settings for training. It resides inside the overall `ConfigContainer`, which represents the complete configuration for a training run. For the complete API and all available fields, see {py:class}`~[megatron.bridge.training.config.ProfilingConfig](https://docs.nvidia.com/nemo/megatron-bridge/latest/apidocs/bridge/bridge.training.config.html#bridge.training.config.ProfilingConfig)`.
```
Comment (maanug-nv) [inline docs/profiling.md]: link formatting seems broken:
image

File: docs/profiling.md [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+
+When using NSys profiling, launch your training script with the NSys command wrapper:
+
+```bash
+nsys profile -s none -o <profile_filepath> -t cuda,nvtx --force-overwrite true --capture-range=cudaProfilerApi --capture-range-end=stop python <path_to_script>
+```
+
+Replace `<profile_filepath>` with your desired output path and `<path_to_script>` with your training script. The `--capture-range=cudaProfilerApi` option ensures profiling is controlled by the framework's step range configuration.
+
+### Configure Profiling with the NSys Plugin
```
Comment (maanug-nv) [inline docs/profiling.md]: nit: i think we can mention NeMo Run in the header itself. if im not using NeMo Run, ill just ignore the section (since command wrapper is already mentioned).
```suggestion
### Configure Profiling with the NeMo Run NSys Plugin
```

File: docs/profiling.md (hunk 18 above)
Comment (jgerh) [reply docs/profiling.md]: Will this work?
```suggestion
`ProfilingConfig` is a dataclass that encapsulates profiling-related settings for training. It resides inside the overall `ConfigContainer`, which represents the complete configuration for a training run. For the complete API and all available fields, see [`ProfilingConfig`](https://docs.nvidia.com/nemo/megatron-bridge/latest/apidocs/bridge/bridge.training.config.html#bridge.training.config.ProfilingConfig).
```

File: docs/profiling.md (hunk 19 above)
Comment (jgerh) [reply docs/profiling.md]: Good idea.

File: docs/profiling.md (hunk 18 above)
Comment (maanug-nv) [reply docs/profiling.md]: I think this works better:
```suggestion
`ProfilingConfig` is a dataclass that encapsulates profiling-related settings for training. It resides inside the overall `ConfigContainer`, which represents the complete configuration for a training run. For the complete API and all available fields, see {py:class}`~bridge.training.config.ProfilingConfig`.
```
